conbench/entities/history.py

```diff
@@ -85,31 +119,188 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
-        .join(Distribution, Distribution.commit_id == Commit.id)
         .filter(
             BenchmarkResult.case_id == case_id,
             BenchmarkResult.context_id == context_id,
             BenchmarkResult.error.is_(None),
             Commit.sha == Commit.fork_point_sha,  # on default branch
             Commit.repository == repo,
             Hardware.hash == hardware_hash,
-            Distribution.case_id == case_id,
-            Distribution.context_id == context_id,
-            Distribution.hardware_hash == hardware_hash,
         )
-        .order_by(Commit.timestamp.asc())
-        .all()
+        .subquery()
+    )
+
+    # Add column with cumulative sum of distribution changes
+    history = Session.query(
+        history,
+        func.sum(history.c.begins_distribution_change.cast(s.Integer))
+        .over(order_by=history.c.timestamp)
+        .label("segment_id"),
+    ).subquery()
+
+    # Add column with rolling mean of the means (only inside of the segment)
+    # Exclude the current commit first...
+    history = Session.query(
+        history,
+        func.coalesce(
+            func.avg(history.c.mean).over(
+                partition_by=history.c.segment_id,
+                order_by=s.extract("epoch", history.c.timestamp),
+                range_=(-Config.ROLLING_WINDOW_SIZE_SEC, -1),
+            ),
+            history.c.mean,
+        ).label("rolling_mean_excluding_this_commit"),
+    ).subquery()
+
+    # ...but if requested, add another column including the current commit
+    if include_current_commit_in_rolling_stats:
+        history = Session.query(
+            history,
+            func.avg(history.c.mean)
+            .over(
+                partition_by=history.c.segment_id,
+                order_by=s.extract("epoch", history.c.timestamp),
+                range_=(-Config.ROLLING_WINDOW_SIZE_SEC, 0),
+            )
+            .label("rolling_mean"),
+        ).subquery()
+    else:
+        history = Session.query(
+            history, history.c.rolling_mean_excluding_this_commit.label("rolling_mean")
+        ).subquery()
+
+    # Add column with the residuals from the exclusive rolling mean, since we always
+    # want to compare to the baseline distribution
+    history = Session.query(
+        history,
+        (history.c.mean - history.c.rolling_mean_excluding_this_commit).label(
+            "residual"
+        ),
+    ).subquery()
+
+    # Add column with the rolling standard deviation of the residuals
+    # (these can go outside the segment since we assume they don't change much)
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(
+            order_by=s.extract("epoch", history.c.timestamp),
+            range_=(
+                -Config.ROLLING_WINDOW_SIZE_SEC,
+                0 if include_current_commit_in_rolling_stats else -1,
+            ),
+        )
+        .label("rolling_stddev"),
     )
+
+    return history.all()
```

Comment:
It might be helpful to render the SQL for these queries here so we can think about ways to adapt this to having a window of fixed commit sizes. Is that something you could do and post here?

Or maybe it would be helpful for the two of us to pair on this tomorrow and see if we can get this calculation going